libyul/backends/evm/EVMDialect.cpp

```diff
@@ -195,112 +179,203 @@ std::set<std::string, std::less<>> createReservedIdentifiers(langutil::EVMVersio
 		std::vector<std::optional<LiteralKind>> _literalArguments,
 		std::function<void(FunctionCall const&, AbstractAssembly&, BuiltinContext&)> _generateCode
-	) -> std::optional<BuiltinFunctionForEVM>
+	)
+{
+	yulAssert(_literalArguments.size() == _params || _literalArguments.empty(), "");
+
+	BuiltinFunctionForEVM f;
+	f.name = _name;
+	f.numParameters = _params;
+	f.numReturns = _returns;
+	f.sideEffects = _sideEffects;
+	f.controlFlowSideEffects = _controlFlowSideEffects;
+	f.literalArguments = std::move(_literalArguments);
+	f.isMSize = false;
+	f.instruction = {};
+	f.generateCode = std::move(_generateCode);
+	return f;
+}
+
+class BuiltinFunctionCollection
+{
+	static size_t constexpr instructionBit = 0;
+	static size_t constexpr replacedInstructionBit = 1;
+	static size_t constexpr objectAccessBit = 2;
+	static size_t constexpr requiresEOFBit = 3;
+	static size_t constexpr requiresNonEOFBit = 4;
+public:
+	struct BuiltinDescription
 	{
-		if (!_objectAccess)
-			return std::nullopt;
-		return createFunction(_name, _params, _returns, _sideEffects, _controlFlowSideEffects, std::move(_literalArguments), std::move(_generateCode));
+		bool instruction() const { return value.test(instructionBit); }
+		bool replaced() const { return value.test(replacedInstructionBit); }
+		bool objectAccess() const { return value.test(objectAccessBit); }
+		bool requiresEOF() const { return value.test(requiresEOFBit); }
+		bool requiresNonEOF() const { return value.test(requiresNonEOFBit); }
+
+		BuiltinDescription operator|(BuiltinDescription const& _other) const
+		{
+			return { value | _other.value };
+		}
+
+		std::bitset<5> value;
 	};
 
-	builtins.emplace_back(createIfObjectAccess("linkersymbol", 1, 1, SideEffects{}, ControlFlowSideEffects{}, {LiteralKind::String}, [](
-		FunctionCall const& _call,
-		AbstractAssembly& _assembly,
-		BuiltinContext&
-	) {
-		yulAssert(_call.arguments.size() == 1, "");
-		Expression const& arg = _call.arguments.front();
-		_assembly.appendLinkerSymbol(formatLiteral(std::get<Literal>(arg)));
-	}));
-	builtins.emplace_back(createIfObjectAccess(
-		"memoryguard",
-		1,
-		1,
-		SideEffects{},
-		ControlFlowSideEffects{},
-		{LiteralKind::Number},
-		[](
-			FunctionCall const& _call,
+	static std::vector<std::tuple<BuiltinDescription, BuiltinFunctionForEVM>> const& builtins()
+	{
+		static BuiltinFunctionCollection const instance;
+		return instance.m_builtinFunctions;
+	}
+
+private:
+	static BuiltinDescription constexpr instruction{1 << instructionBit};
+	static BuiltinDescription constexpr replaced{1 << replacedInstructionBit};
+	static BuiltinDescription constexpr objectAccess{1 << objectAccessBit};
+	static BuiltinDescription constexpr requiresEOF{1 << requiresEOFBit};
+	static BuiltinDescription constexpr requiresNonEOF{1 << requiresNonEOFBit};
+
+	BuiltinFunctionCollection()
+	{
+		for (auto const& [name, opcode]: evmasm::c_instructions)
+		{
+			if (isLowLevelStackManipulationInstruction(opcode))
+				continue;
```

Comment:
I have to skip them already here because there are asserts in the side-effect gathering that prevent me from creating builtin functions for these instructions in the first place. We could of course remove these asserts. Or just skip them here, before creating anything. :)